src/array_api_extra/_lib/_quantile.py

```diff
@@ -0,0 +1,149 @@
+"""Quantile implementation."""
+
+from types import ModuleType
+from typing import cast
+
+from ._at import at
+from ._utils import _compat
+from ._utils._compat import array_namespace
+from ._utils._typing import Array
+
+
+def quantile(
+    x: Array,
+    q: Array | float,
+    /,
+    *,
+    axis: int | None = None,
+    keepdims: bool | None = None,
+    method: str = "linear",
+    xp: ModuleType | None = None,
+) -> Array:  # numpydoc ignore=PR01,RT01
+    """See docstring in `array_api_extra._delegation.py`."""
+    if xp is None:
+        xp = array_namespace(x, q)
+
+    q_is_scalar = isinstance(q, int | float)
+    if q_is_scalar:
+        q = xp.asarray(q, dtype=xp.float64, device=_compat.device(x))
+    q_arr = cast(Array, q)
+
+    if not xp.isdtype(x.dtype, ("integral", "real floating")):
+        msg = "`x` must have real dtype."
+        raise ValueError(msg)
+    if not xp.isdtype(q_arr.dtype, "real floating"):
+        msg = "`q` must have real floating dtype."
+        raise ValueError(msg)
+
+    # Promote to common dtype
+    x = xp.astype(x, xp.float64)
+    q_arr = xp.asarray(q_arr, dtype=xp.float64, device=_compat.device(x))
+
+    dtype = x.dtype
+    axis_none = axis is None
+    ndim = max(x.ndim, q_arr.ndim)
+
+    if axis_none:
+        x = xp.reshape(x, (-1,))
+        q_arr = xp.reshape(q_arr, (-1,))
+        axis = 0
+    elif not isinstance(axis, int):  # pyright: ignore[reportUnnecessaryIsInstance]
+        msg = "`axis` must be an integer or None."
+        raise ValueError(msg)
+    elif axis >= ndim or axis < -ndim:
+        msg = "`axis` is not compatible with the shapes of the inputs."
+        raise ValueError(msg)
+    else:
+        axis = int(axis)
+
+    if keepdims not in {None, True, False}:
+        msg = "If specified, `keepdims` must be True or False."
+        raise ValueError(msg)
+
+    if x.shape[axis] == 0:
+        shape = list(x.shape)
+        shape[axis] = 1
+        x = xp.full(shape, xp.nan, dtype=dtype, device=_compat.device(x))
+
+    y = xp.sort(x, axis=axis)
```

Comment:
Note: `xp.sort` is stable by default, and this sort doesn't need to be stable. Stable sort is much slower.

See this issue: https://github.com/scikit-learn/scikit-learn/issues/32283

```suggestion
    y = xp.sort(x, axis=axis, stable=False)
```